zebrad/src/components/mempool/tests.rs

```diff
@@ -219,3 +221,125 @@ async fn mempool_queue() -> Result<(), Report> {
+        zebra_state::init(state_config.clone(), network);
+
+    let mut state_service = ServiceBuilder::new().buffer(1).service(state);
+    let (_chain_verifier, tx_verifier) =
+        zebra_consensus::chain::init(consensus_config.clone(), network, state_service.clone())
+            .await;
+
+    // Start the mempool service
+    let mut mempool = Mempool::new(
+        network,
+        Buffer::new(BoxService::new(peer_set), 1),
+        state_service.clone(),
+        tx_verifier,
+        sync_status,
+        chain_tip_change,
+    );
+
+    // Push the genesis block to the state
+    let genesis_block: Arc<Block> = zebra_test::vectors::BLOCK_MAINNET_GENESIS_BYTES
+        .zcash_deserialize_into()
+        .unwrap();
+    state_service
+        .ready_and()
+        .await
+        .unwrap()
+        .call(zebra_state::Request::CommitFinalizedBlock(
+            genesis_block.clone().into(),
+        ))
+        .await
+        .unwrap();
+
+    // Queue transaction from block 2 for download
+    let txid = block2.transactions[0].unmined_id();
+    let response = mempool
+        .ready_and()
+        .await
+        .unwrap()
+        .call(Request::Queue(vec![txid.into()]))
+        .await
+        .unwrap();
+    let queued_responses = match response {
+        Response::Queued(queue_responses) => queue_responses,
+        _ => unreachable!("will never happen in this test"),
+    };
+    assert_eq!(queued_responses.len(), 1);
+    assert!(queued_responses[0].is_ok());
+    assert_eq!(mempool.tx_downloads.in_flight(), 1);
+
+    // Query the mempool to make it poll chain_tip_change
+    for _ in 0..10 {
+        let _response = mempool
+            .ready_and()
+            .await
+            .unwrap()
+            .call(Request::TransactionIds)
+            .await
+            .unwrap();
+    }
+
+    // Push block 1 to the state
+    println!("Comitting block 1...");
+    state_service
+        .ready_and()
+        .await
+        .unwrap()
+        .call(zebra_state::Request::CommitFinalizedBlock(
+            block1.clone().into(),
+        ))
+        .await
+        .unwrap();
+
+    // Query the mempool to make it poll chain_tip_change
+    let _response = mempool
+        .ready_and()
+        .await
+        .unwrap()
+        .call(Request::TransactionIds)
+        .await
+        .unwrap();
+
+    // Push block 2 to the state
+    println!("Comitting block 2...");
+    state_service
+        .oneshot(zebra_state::Request::CommitFinalizedBlock(
+            block2.clone().into(),
+        ))
+        .await
+        .unwrap();
+
+    for _ in 0..10 {
+        // Query the mempool just to poll it and make it cancel the download.
+        let _response = mempool
+            .ready_and()
+            .await
+            .unwrap()
+            .call(Request::TransactionIds)
+            .await
+            .unwrap();
+    }
+
+    // Check if download was cancelled.
+    assert_eq!(mempool.tx_downloads.in_flight(), 0);
```

Comment:
I'm stuck with the test (`mempool_cancel_mined`). I'm not sure why it doesn't pass. Could you please take a look @teor2345 (maybe @jvff too?)

I've littered the code with some prints to debug this. I've comitted them just to make it easier to explain and I'll remove them later.

What basically happens is that, when block 2 is comitted, the mempool does receive a `TipAction::Grow` and it does cancel the download, i.e. this runs:


https://github.com/ZcashFoundation/zebra/blob/3239f9c67133113a2176b95b8416479695f17b4d/zebrad/src/components/mempool/downloads.rs#L341-L343

However, the cancel does not seem to go through, i.e. this select does not return:

https://github.com/ZcashFoundation/zebra/blob/3239f9c67133113a2176b95b8416479695f17b4d/zebrad/src/components/mempool/downloads.rs#L299-L306

And from the prints I can see that the downloader is being polled, i.e. this first line prints but it always returns `Pending` in `ready!`

https://github.com/ZcashFoundation/zebra/blob/3239f9c67133113a2176b95b8416479695f17b4d/zebrad/src/components/mempool/downloads.rs#L152-L156

Could this be related that the task should be blocked in 

https://github.com/ZcashFoundation/zebra/blob/3239f9c67133113a2176b95b8416479695f17b4d/zebrad/src/components/mempool/downloads.rs#L257-L258

trying to download using the mock network service (which I think it's stuck in a timeout, since I haven't specifically mocked the request/response that should happen)? Will select only return when that unblocks?

Reply:
I think I found the issue. My guess is that it's a starvation issue. I.e., no `await` point in the test task ever returns `Poll::Pending`, so it runs to the end as fast as possible, not giving a chance for the background task to complete the cancellation. I added a `tokio::task::yield_now().await;` inside the last `for` loop, and the test now seems to pass. If I reduce the iterations of the `for` loop to `1`, then the test starts failing again. Setting it to `2` and above makes it succeed.

Reply:
I'm not sure this helps, but one idea to try to have this a bit more deterministic is to use `tokio::time::pause()` and `tokio::time::sleep(some_time).await`. It might sound counter intuitive, but the idea is that the sleep will actually only run while there is work to do on the executor, as soon as it becomes idle, it advances time up to the moment the sleep wakes up, so it kind of runs ASAP.

So every time you need wait for something to happen, just sleep for a time (it could be any time AFAIK, even something like an hour, it will still wake up as soon as possible, but it would be best to add something that doesn't cause surprises later, so maybe a few hundred milliseconds).